PLOS/wombat
# Ambr 1035

https://jira.plos.org/jira/browse/AMBR-1035

## To test:

- build companion PR and install to apt repo https://github.com/PLOS/rhino/pull/503: `mvn install`
- build this PR: `mvn install`
- bring up ambra VM:
  ```
  $ cd pogos/salt/molten/vagrant/apps/ambra
  $ vagrant up
  ```
- `ssh` into `ambra` and make some edits to get a full set of relationship data:
- Edit rhino db setup:
```
egh@ambra:~$ sudo vi /opt/plos/rhino/conf/context.xml 
```
Change the db `Resource` to:
```xml
 <Resource name="jdbc/AmbraDS"
              auth="Container"
              type="javax.sql.DataSource"
              factory="org.apache.tomcat.jdbc.pool.DataSourceFactory"
              validationQuery="SELECT 1"
              testOnBorrow="true"
              driverClassName="com.mysql.jdbc.Driver"
              username="ambra"
              password="XXXXXX"
              url="jdbc:mysql://db-ambra-dev.soma.plos.org:3306/ambra"
              maxActive="300"/>
```
(check with me for the password or `sudo cat /opt/plos/rhino/conf/context.xml` on `journals-dev1-backend-soma.plos.org`
-  Fix hosts: (needed to access dev crepo):
```
egh@ambra:~$ echo 10.5.4.5 dev-main.journals-contentrepo.service.consul | sudo tee -a /etc/hosts
```
-  Point rhino to dev crepo:
```
egh@ambra:~$ sudo vi /opt/plos/rhino/conf/rhino.yaml 
```
Change all `address` entries to: `http://dev-main.journals-contentrepo.service.consul:8002/v1/`
- Restart rhino:
```
egh@ambra:~$ sudo service rhino restart
```

- Visit some pages with interesting related articles:
- http://journals.vagrant.local/plosbiology/article?id=10.1371/journal.pbio.1002593
- http://journals.vagrant.local/plosone/article?id=10.1371/journal.pone.0160700

## Notes

In ended up rewriting a lot of the related article code to make it simpler and better modeled.

In rhino, we now normalize all relations to be “inbound”. All article relationships consist of a “source”, a “target” and a “type”. (This is a directed labeled graph.) Each type also has a related, inverted direction versions, e.g. a `corrected-article` has `correction-forward`. Rhino was previously reporting some links about a certain article in two directions, “inbound” and “outbound”. This was confusing because some relations would be listed in one direction but not the other, and sometimes both. This was fine when we just listed relationships on the sidebar, but now that we want to properly label their type and “direction” (is this article a correction of another article, or is it corrected by another article?) the situation was untenable.

Here is an example of how the API changes:
Old:
```
{
  "inbound": [
    {
      "type": "corrected-article",
      "doi": "10.1371/journal.pone.0163840",
      "revisionNumber": 1,
      "title": "\u003carticle-title xmlns:mml\u003d\"http://www.w3.org/1998/Math/MathML\" xmlns:xlink\u003d\"http://www.w3.org/1999/xlink\"\u003eCorrection: HIF-2α Expression Regulates Sprout Formation into 3D Fibrin Matrices in Prolonged Hypoxia in Human Microvascular Endothelial Cells\u003c/article-title\u003e",
      "publicationDate": "2016-09-22",
      "journal": {
        "journalKey": "PLoSONE",
        "title": "PLOS ONE",
        "eIssn": "1932-6203"
      }
    }
  ],
  "outbound": [
    {
      "type": "correction-forward",
      "doi": "10.1371/journal.pone.0163840",
      "revisionNumber": 1,
      "title": "\u003carticle-title xmlns:mml\u003d\"http://www.w3.org/1998/Math/MathML\" xmlns:xlink\u003d\"http://www.w3.org/1999/xlink\"\u003eCorrection: HIF-2α Expression Regulates Sprout Formation into 3D Fibrin Matrices in Prolonged Hypoxia in Human Microvascular Endothelial Cells\u003c/article-title\u003e",
      "publicationDate": "2016-09-22",
      "journal": {
        "journalKey": "PLoSONE",
        "title": "PLOS ONE",
        "eIssn": "1932-6203"
      }
    }
  ]
}
```

New:
```
[
  {
    "doi": "10.1371/journal.pone.0163840",
    "publicationDate": "2016-09-22",
    "revisionNumber": 1,
    "title": "\u003carticle-title xmlns:mml\u003d\"http://www.w3.org/1998/Math/MathML\" xmlns:xlink\u003d\"http://www.w3.org/1999/xlink\"\u003eCorrection: HIF-2α Expression Regulates Sprout Formation into 3D Fibrin Matrices in Prolonged Hypoxia in Human Microvascular Endothelial Cells\u003c/article-title\u003e",
    "type": "corrected-article",
    "journal": {
      "journalKey": "PLoSONE",
      "title": "PLOS ONE",
      "eIssn": "1932-6203"
    }
  }
]
```

Other changes include:
- rhino/wombat: using google autovalue for these model classes to simplify somewhat
- wombat: using classes to model this data, rather than `Map` objects
- wombat: consolidating all code that hits the `relationships` endpoint in rhino, rather than having two different places that is called/parsed
